### PR TITLE
JSON request/response mixin

### PR DESCRIPTION
**Pull request by untitaker**
Continuing #1416

All inline discussions have been resolved.

### CODE REVIEW

**Comment by untitaker** [comment]
@homu r+

File: flask/wrappers.py [hunk 1]
```diff
@@ -13,19 +13,110 @@
...
+        if not (force or self.is_json):
+            return None
+
+        # We accept MIME charset header against the specification as certain
+        # clients have been using this in the past.  For responses, we assume
+        # that if the response charset was set explicitly then the data had
+        # been encoded correctly as well.
+        charset = self.mimetype_params.get('charset')
+        try:
+            data = self._get_data_for_json(cache)
```
Comment (untitaker) [inline flask/wrappers.py]: Do we even want to pass `cache` on? If we do this, by default both `data` and `json` are stored on the request object. While this leads to the least surprise for newcomers (no empty return values), it is very memory inefficient.

File: flask/wrappers.py [hunk 2]
```diff
@@ -13,19 +13,110 @@
...
+        """
+        from warnings import warn
+        warn(DeprecationWarning(
+            'json is deprecated.  Use get_json() instead.'), stacklevel=2)
+        return self.get_json()
+
+    def _get_data_for_json(self, cache):
+        getter = getattr(self, 'get_data', None)
+        if getter is not None:
+            return getter(cache=cache)
```
Comment (untitaker) [inline flask/wrappers.py]: We have a problem here: `werkzeug.Response.get_data` doesn't accept a `cache` flag.

Comment (mitsuhiko) [inline flask/wrappers.py]: Can we just add the flag there and require a newer werkzeug?

Comment (untitaker) [inline flask/wrappers.py]: Yes, see https://github.com/pallets/werkzeug/pull/992

**Comment by sevankj** [comment]
Reviewed 5 of 7 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pallets/flask/1984)*

**Comment by JordanP** [comment]
Any progress on this ? Could you rebase ?

I've been looking for better JSON support (a-la python-requests) and there's already several bug reports and PR in flask and werkzeug about this. This looks like a popular demand :)

**Comment by jwg4** [comment]
This change would make a big difference to us. Sorry to see that it is nearly two years old, can I do something to help?

File: flask/wrappers.py (hunk 1 above)
Comment (JordanP) [inline flask/wrappers.py]: The `cache` arg is already defined on the original `get_json()` method, so I don't think removing it or changing its behavior should be part of this PR.

**Comment by leejunhoan** [comment]
Reviewed 5 of 7 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pallets/flask/1984)*

**Comment by ramnes** [comment]
This request merged would be a huge relief for me and my team also.

@mitsuhiko, @untitaker, how can we make this happen?

**Comment by kislyuk** [comment]
+1 for this feature. Any way to help this get in?

**Comment by kennethreitz** [comment]
needs a rebase!

**Comment by kennethreitz** [comment]
Maybe it would be better to use Requests as the test client :)

File: flask/wrappers.py [hunk 3]
```diff
@@ -13,19 +13,110 @@
...
+
+    @property
+    def json(self):
+        """If this request/response is in JSON format then this property will
+        contain the parsed JSON data.  Otherwise it will be ``None``.
+
+        The :meth:`get_json` method should be used instead.
+        """
+        from warnings import warn
+        warn(DeprecationWarning(
```
Comment (kennethreitz) [inline flask/wrappers.py]: `.json` is important, and shouldn't be deprecated

Comment (kennethreitz) [inline flask/wrappers.py]: `is_json()` shouldn't be neccessary, as `.json` should return `None` if there is no JSON, correct?

**Comment by kennethreitz** [comment]
closing, old pr, requests is actually a better idea for the client (though i'm sure that's not an obvious choice)

**Comment by untitaker** [comment]
That is only your opinion at the moment, and the Flask docs don't show how to
use it as one. Before we settle on this we should make clear how to actually
use it in tests.

Right now you have to spin up a dev server for this.

**Comment by kennethreitz** [comment]
i definitely don't think requests should go in, i'm mostly joking about that :)

**Comment by kennethreitz** [comment]
may have been premature to close this — i reviewed every line of code, and it seemed solid until i got to the deprecation of `.json`. then, I saw that there may be some issues with werkzeug and saw the merge conflicts so decided to close.

**Comment by untitaker** [comment]
I'm not sure either what to think of that.

We can have (parts of) requests as test client and it would probably be an UX
improvement, but right now I consider the overhead of its entire HTTP stack to
be too much.

**Comment by kennethreitz** [comment]
Well, you'd write a connection adapter that would skip it speaking HTTP and have it talk WSGI instead. Should be <30LOC. It's alright though, people are used to having unique test clients, and people are def used to the one built in here.

**Comment by davidism** [comment]
Continued in #2358.

**Pull request closed without merging**